src/dom/cached-element-vector.ts

```diff
@@ -0,0 +1,157 @@
+ * if they are being used elsewhere.
+ */
+const uses: DynamicDefaultMap<CachedElementVector<MultiDimensionalVector>, Set<any>> =
+    DynamicDefaultMap.usingFunction(() => new Set());
+
+export abstract class CachedElementVector<T extends MultiDimensionalVector> {
+  /**
+   * Used when a single type of caching could serve multiple elements.
+   * For instance you may want to cache the scroll of the main document but also
+   * the scroll of an inner scrollable element.
+   */
+  static getForElement(use: any, args: any[] = [null]): any {
+    const instance = caches.get(this).get(args);
+    uses.get(instance).add(use);
+    return instance;
+  }
+
+  /**
+   * Returns a generic singleton.
+   */
+  static getSingleton(use: any): any {
+    return this.getForElement(use);
+  }
+
+  /**
+   * Type of vector this is caching, Dimensions/Vector2d/etc.
+   */
+  protected static VectorClass: typeof MultiDimensionalVector =
+      MultiDimensionalVector;
+
+  // Element whose values are being cached.
+  protected element: HTMLElement;
+
+  // Optional args the constructor was called with.
+  private readonly args: any[];
+
+  // Currently cached values
+  private values: T[];
+
+  // Allows caches to persist for a little bit so they aren't being created
+  // and destroyed every frame by short-lived calls.
+  private disposeTimeout: number;
+
+  private raf: Raf;
+
+  protected constructor(element: any = null, ...args: any[]) {
+    const instanceByElement = caches.get(this.constructor);
+    this.raf = new Raf(() => this.loop());
+    this.args = [element, ...args];
+
+    // Prevent direct instantiation
+    if (instanceByElement.has([element, ...args])) {
+      if (element) {
+        throw new Error('Please use getForElement instead of new.');
+      } else {
+        throw new Error('Please use getSingleton instead of new.');
+      }
+    }
+
+    this.disposeTimeout = null;
+    this.element = element;
+    this.values = <T[]>[this.getCurrentVector()];
+    this.init();
+  }
+
+  getLastValue(): T {
+    return this.values.slice(-1)[0];
+  }
+
+  getDelta(): T {
+    const values = this.getCurrentAndLastValue();
+    return <T>this.getVectorClass().subtract(values[0], values[1]);
+  }
+
+  hasChanged(): boolean {
+    return !this.getVectorClass().areEqual(...this.getCurrentAndLastValue());
+  }
+
+  destroy(use: any): void {
+    uses.get(this).delete(use);
+    clearTimeout(this.disposeTimeout);
+    // Hang tight for a minute so we aren't creating and destroying these caches
+    // too rapidly.
+    this.disposeTimeout = window.setTimeout(
+        () => {
+          if (uses.size <= 0) {
+            caches.get(this.constructor).delete(this.args);
+            this.raf.dispose();
+          }
+        },
+        1000);
+  }
+
+  // Expected to be overridden
+  protected getValues(): number[] {
+    throw new Error('getValues must be overridden by child class');
+  }
+
+  private init(): void {
+    this.raf.start();
```

Comment:
Make sure we dispose raf.

Reply:
It gets disposed on line 118 in destroy()

But that reminded me of another thing I meant to do.

I renamed `destroy()` to `dispose()` to align w/ yano convention.